sdk/Pulumi/Converter/Converter.cs

```diff
@@ -0,0 +1,259 @@
+using System;
+using System.Net;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using System.Linq;
+using System.Collections.Generic;
+using Grpc.Core;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Hosting;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Hosting.Server.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+using Pulumirpc.Codegen;
+using Diagnostic = Pulumi.Codegen.Diagnostic;
+using DiagnosticSeverity = Pulumi.Codegen.DiagnosticSeverity;
+
+namespace Pulumi.Experimental.Converter
+{
+    public sealed class ConvertProgramRequest
+    {
+        public readonly string SourceDirectory;
+        public readonly string TargetDirectory;
+        public readonly string MapperTarget;
+
+        public ConvertProgramRequest(string sourceDirectory, string targetDirectory, string mapperTarget)
+        {
+            SourceDirectory = sourceDirectory;
+            TargetDirectory = targetDirectory;
+            MapperTarget = mapperTarget;
+        }
+    }
+
+    public sealed class ConvertProgramResponse
+    {
+        public IList<Diagnostic>? Diagnostics { get; set; }
+        public static ConvertProgramResponse Empty => new ConvertProgramResponse();
+    }
+
+    public abstract class Converter
+    {
+        public virtual Task<ConvertProgramResponse> ConvertProgram(ConvertProgramRequest request)
+        {
+            throw new NotImplementedException($"{nameof(ConvertProgram)} is not implemented");
+        }
+
+        public static Converter CreateSimple(Func<ConvertProgramRequest, ConvertProgramResponse> convertProgram)
+        {
+            return new SimpleProgramConverter(request =>
+            {
+                var response = convertProgram(request);
+                return Task.FromResult(response);
+            });
+        }
+
+        public static Converter CreateSimpleAsync(Func<ConvertProgramRequest, Task<ConvertProgramResponse>> convertProgramAsync)
+        {
+            return new SimpleProgramConverter(convertProgramAsync);
+        }
+
+        public static async Task Serve(
+            string[] args,
+            Converter converter,
+            CancellationToken? cancellationToken = null,
+            TextWriter? stdout = null)
+        {
+            // maxRpcMessageSize raises the gRPC Max message size from `4194304` (4mb) to `419430400` (400mb)
+            var maxRpcMessageSize = 400 * 1024 * 1024;
+
+            var host = Host.CreateDefaultBuilder()
```

Comment:
We ought to find a way to share this between Converters and Providers. Every plugin is going to look like this in the end.

Reply:
Some parts can be shared, especially where we get the port and print it out. I was wondering though, does the engine send any args to converters plugins? I am thinking this piece of code is redundant:
```cs
var memConfig = new Dictionary<string, string>();
if (args.Length > 0)
{
    memConfig.Add("Host", args[0]);
}
config.AddInMemoryCollection(memConfig);
```
Which could also mean that we can remove the `string[] args` parameter from `Serve`

Reply:
> does the engine send any args to converters plugins?

Not at the moment.

Reply:
Nice, so it can be removed from the API 😄 if we decide that it is needed later on, we can just call `Environment.GetCommandLineArgs()`